src/services/redisUserMetadataStorage.ts

```diff
@@ -0,0 +1,105 @@
+import * as redis from "redis";
+
+import { Either, isLeft, isRight, left, right } from "fp-ts/lib/Either";
+import { ReadableReporter } from "italia-ts-commons/lib/reporters";
+import { User } from "src/types/user";
+import { UserMetadata } from "../../generated/backend/UserMetadata";
+import { log } from "../utils/logger";
+import { IUserMetadataStorage } from "./IUserMetadataStorage";
+import RedisStorageUtils from "./redisStorageUtils";
+
+const userMetadataPrefix = "USERMETA-";
+export const metadataNotFoundError = new Error("User Metadata not found");
+export const invalidVersionNumberError = new Error("Invalid version number");
+
+export default class RedisUserMetadataStorage extends RedisStorageUtils
+  implements IUserMetadataStorage {
+  constructor(private readonly redisClient: redis.RedisClient) {
+    super();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public async set(
+    user: User,
+    payload: UserMetadata
+  ): Promise<Either<Error, boolean>> {
+    const getUserMetadataResult = await this.loadUserMetadataByFiscalCode(
+      user.fiscal_code
+    );
+    if (
+      isRight(getUserMetadataResult) &&
+      getUserMetadataResult.value.version !== payload.version - 1
+    ) {
+      return left(invalidVersionNumberError);
+    }
+    if (
+      isLeft(getUserMetadataResult) &&
+      getUserMetadataResult.value !== metadataNotFoundError
+    ) {
+      return left(getUserMetadataResult.value);
+    }
+    return await new Promise<Either<Error, boolean>>(resolve => {
```

Comment:
this is not the correct semantic, if two clients attempt to set a certain version concurrently, they will both succeed

you need to implement [compare-and-set](https://redis.io/topics/transactions#cas)

Reply:
I understand, but a solution is not so simple. The value stored on redis is not a simple integer value, but a stringified JSON having the property `version: number`. Nevertheless the specific API call works on a resource shared only with devices of the same user, concurrency on storage with different metadata values is a rare scenario.

Reply:
nobody said it was simple :D

ok so add a BIG comment explaining that the operation is not safe for concurrent updates, also create a story in the backend backlog for implementing proper CAS and link to that story from the comment here

Reply:
Fix will be implemented in [this](https://www.pivotaltracker.com/story/show/167064659) story

Reply:
Code comments added in commit fc65a2a